### PR TITLE
Phase 6: Gradient Centralization — Zero-Mean Gradient Updates with Lion

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Lion optimizer uses sign(gradient_EMA) for weight updates — discarding all gradient magnitude information. This means the model's update direction at each step is determined entirely by the sign of the mean gradient vector. If a weight matrix has gradients with a large DC offset (all components pointing in the same direction), Lion's sign operation is dominated by this shared direction, reducing effective gradient diversity.

**Gradient Centralization (GC)** (Yong et al., ECCV 2020) removes the DC component of gradients before the optimizer step. For a weight matrix W with gradient ∇W, GC computes:

```
∇W_centered = ∇W - mean(∇W, dim=[1, 2, ..., k-1], keepdim=True)
```

where the mean is taken over all dimensions except the first (output channels). This constrains the gradient update to be zero-mean across input dimensions.

**Why this helps with Lion specifically:** Lion uses sign(EMA(∇W_centered)) instead of sign(EMA(∇W)). By centering, we ensure that Lion's binary updates are more balanced across output neurons — no single direction dominates. This is especially important for our multi-task loss (volume + surface + surface_refine + re_head + aoa_head) where gradients from different task heads can create large DC components that bias the optimizer.

**Why this might help p_tan:** The tandem distribution is a small fraction of training data. Its gradient contributions may be systematically suppressed by the large DC offset from single-foil gradients. GC removes this bias, giving tandem-relevant gradient directions equal representation in Lion's sign operation.

**Novel interaction:** GC was originally validated with SGD and Adam. Its interaction with Lion's sign operation is unstudied — this is a genuinely novel combination. Zero extra parameters. ~5 lines. Zero inference cost.

**Paper:** Yong et al., "Gradient Centralization: A New Optimization Technique for Deep Neural Networks", ECCV 2020. Showed consistent 0.5-3% improvements on ImageNet with SGD/Adam.

## Instructions

**Step 1: Add flag to argparse:**
```python
parser.add_argument('--gradient_centralization', action='store_true', default=False,
                    help='Apply gradient centralization before optimizer step (GC-Lion)')
```
Add `gradient_centralization: bool = False` to Config dataclass.

**Step 2: Add GC function (top of file, near other utility functions):**
```python
def apply_gradient_centralization(model: nn.Module) -> None:
    """Subtract gradient mean over input dims (GC, Yong et al. ECCV 2020)."""
    for p in model.parameters():
        if p.grad is not None and p.grad.dim() > 1:
            axes = tuple(range(1, p.grad.dim()))
            p.grad.data -= p.grad.data.mean(dim=axes, keepdim=True)
```

**Step 3: Insert AFTER `loss.backward()` and BEFORE `torch.nn.utils.clip_grad_norm_` in the main training loop** (around line 1769-1771):

```python
        loss.backward()

        if cfg.gradient_centralization:
            apply_gradient_centralization(model)
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
```

No other changes needed. GC applies to all weight tensors (dim > 1), skipping biases (dim=1).

**Note on PCGrad:** `--disable_pcgrad` is set in all runs below, so the PCGrad interaction is not relevant here.

**Experiment runs (2 seeds):**

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_name "tanjiro/grad-cent-s42" \
  --wandb_group phase6/gradient-centralization \
  --gradient_centralization --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

# Seed 73
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_name "tanjiro/grad-cent-s73" \
  --wandb_group phase6/gradient-centralization \
  --gradient_centralization --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

W&B group: `phase6/gradient-centralization`

**What to report:**
- Surface MAE (p_in, p_oodc, p_tan, p_re) for each run — all 4 required, no NaN accepted
- val/loss at convergence
- W&B run IDs

## Baseline

Current best (16-seed ensemble, PR #2093):

| Metric | Ensemble Baseline | Single-model 8-seed mean |
|--------|-------------------|--------------------------|
| p_in   | **12.1**          | 13.03 |
| p_oodc | **6.6**           | 7.83  |
| p_tan  | **29.1**          | 30.29 |
| p_re   | **5.8**           | 6.45  |

Single-model references (same seed comparison):
- seed=42: p_in≈12.71, p_oodc≈8.13
- seed=73: p_in≈12.2, p_oodc≈7.59

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_name "tanjiro/baseline-s42" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### Surface Pressure MAE (lower is better)

| Run | p_in | p_oodc | p_tan | p_re | val_loss | W&B |
|-----|------|--------|-------|------|----------|-----|
| baseline-s42 (from #2101) | 13.3 | 7.7 | 30.3 | 6.5 | 0.388 | `tqlbfz9y` |
| baseline-s73 (from #2101) | 12.9 | 8.1 | 30.5 | 6.6 | 0.392 | `dck4ur8w` |
| **grad-cent-s42** | **15.6** | 10.6 | 30.8 | 8.3 | 0.460 | `waad26c0` |
| **grad-cent-s73** | 14.5 | 9.3 | **29.6** | 7.2 | 0.422 | `a2p1auev` |

**Single-model baseline mean (PR body):** p_in=13.03, p_oodc=7.83, p_tan=30.29, p_re=6.45

### Analysis

**Gradient Centralization clearly hurt performance** — both runs are significantly worse than their baseline counterparts. The hypothesis that GC would help Lion by removing gradient DC offset was not supported.

Key observations:

1. **p_in regressed sharply.** s42: 15.6 (+2.3 vs baseline), s73: 14.5 (+1.6). This is the strongest regression seen across all Phase 6 experiments. GC actively disrupted in-distribution learning.

2. **p_oodc and p_re also regressed.** s42: 10.6 p_oodc (vs 7.7 baseline), 8.3 p_re (vs 6.5). s73 was closer but still regressed: 9.3 vs 8.1, 7.2 vs 6.6. 

3. **p_tan inconclusive.** s73 shows p_tan=29.6 (slightly better than 30.5 baseline for that seed), but s42 shows p_tan=30.8 (worse). Variance between seeds is high enough that this is noise, not signal.

4. **val_loss notably higher.** s42: 0.460 vs 0.388, s73: 0.422 vs 0.392. GC significantly disrupted the overall optimization trajectory.

5. **Strong seed dependence.** The gap between s42 and s73 is much larger than usual (p_in: 15.6 vs 14.5, val_loss: 0.460 vs 0.422). This suggests GC destabilizes training, producing high variance outcomes.

**Why GC hurt Lion:** The Lion optimizer's sign operation already discards gradient magnitude — it's inherently direction-only. GC changes the gradient *direction* by removing the mean component. This is beneficial with Adam/SGD where the mean dominates via scale effects, but with Lion the direction is already the only thing that matters. GC essentially introduces noise into Lion's already-binary direction decisions, particularly in the early training layers where the gradient mean carries real signal about the dominant optimization direction.

### Commands Used

```bash
# Seed 42 (GPU 0)
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_group phase6/gradient-centralization --wandb_name "tanjiro/grad-cent-s42" \
  --gradient_centralization --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

# Seed 73 (GPU 1) — same flags
```

Peak memory: ~52 GB (same as baseline)
All runs: 157 epochs, hit 180-min wall-clock timeout.

### Suggested follow-ups

1. **GC with Adam/AdamW instead of Lion** — GC was originally validated with these optimizers. If we ever test without Lion, GC could be worth revisiting.
2. **Partial GC** — apply GC only to certain layers (e.g., output heads or surface refinement head) where gradient bias may be more harmful. Applying GC to all layers may over-constrain the optimization.
3. **Gradient clipping before GC** — current order is backward → GC → clip_grad_norm. Reversing to backward → clip → GC might be more stable.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2114 because DEAD END: Gradient Centralization clearly hurts Lion across all metrics. Both seeds show substantial regression: p_in +12-17%, p_oodc +15-38%, p_re +9-28% vs baseline. Only p_tan on s73 (29.6 vs 30.5) looked potentially interesting, but val/loss deteriorated significantly (0.460 vs 0.388 for s42) suggesting this is noise from a destabilized training run. The student's analysis is correct: Lion's sign(EMA) operation already discards gradient magnitude, so zeroing the mean component changes effective update directions without the magnitude-balancing benefit GC gives Adam/SGD. Gradient direction diversity via zero-mean is counterproductive when sign quantization already enforces binary direction. This direction is closed..
